[Jetpack Focus] Implement reader saved posts ContentProvider and Resolver

develric opened the pull request:
Fixes #17289

To test:

1 - Install WordPress and log in; in the Reader save some posts for later and check they are listed in the SAVED tab
2 - Set the flag isFeatureFlagEnabled to true on SharedLoginResolver, UserFlagsResolver and ReaderSavedPostsResolver (this cannot be done using the Debug settings screen because that is only available when the user is already logged in);
2 - Install Jetpack using the same variant (e.g. if installed WordPress version was jalapenoDebug, install Jetpack jalapenoDebug);
3 - Jetpack should login automatically after launching, and home screen should be shown;
4 - Check the expected reader saved posts are in Jetpack under SAVED tab.
5 - Check the above is working with empty saved posts list scenario
6 - Check the track events are fired.

## Regression Notes
1. Potential unintended areas of impact
    None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
    Manual testing and unit tests

3. What automated tests I added (or what prevented me from doing so)
    Adding unit test in a dedicated PR

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

develric committed:
Adding analytics stats

develric committed:
Adding AppPrefs and wrapper functions

develric committed:
Adding feature flag

develric committed:
Adding needed wrapper and wrapper functions

develric committed:
Adding tracker

develric committed:
Adding provider

develric committed:
Adding resolver

develric committed:
Cascading resolver

develric committed:
Adding user flag

develric committed:
Merge branch 'trunk' into issue/reader-saved-posts-sync

# Resolved Conflicts:
#	WordPress/src/main/java/org/wordpress/android/modules/AppComponent.java
#	WordPress/src/wordpress/AndroidManifest.xml

develric committed:
Fixing detekt

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/reader/savedposts/resolver/ReaderSavedPostsResolver.kt:
```diff
@@ -0,0 +1,102 @@
+                updateReaderSavedPosts(onSuccess, onFailure, posts)
+            } else {
+                readerSavedPostsAnalyticsTracker.trackFailed(ErrorType.NoUserSavedPostsError)
+                onFailure
```
Should we invoke the function here instead, like `onFailure()`?

develric replied:
Of course 🤦‍♂️! Changed [here](206ef3efbb7e49d4167d6c95b9eff50d3f47757f). On a side note, I was working yesterday on the PR that adds unit testing to this code, and a failing test was spotting it for me 😄! Unit testing FTW 🙌!

develric committed:
Fixing unit testing

develric committed:
Fixing function invocation.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/reader/savedposts/resolver/ReaderSavedPostsResolver.kt:
```diff
@@ -0,0 +1,102 @@
+                    )
+            )
+
+            readerTagTableWrapper.getBookmarkTags()!!.first().let {
```
Would it make sense to instead of using the non-null assertion operator to throw an Exception explicitly?
e.g.
```
    readerTagTableWrapper.getBookmarkTags()?.let {
        readerPostTableWrapper.addOrUpdatePosts(it.first(), posts)
    } ?: throw NullPointerException("${ReaderTagList::class.java.simpleName} not found.")

```

develric replied:
Agree in not using the non-null assertion 👍 . Proposing to use the requireNotNulll, let me know 🙇

RenanLukas replied:
Makes sense to me 👍

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/reader/savedposts/provider/ReaderSavedPostsProvider.kt:
```diff
@@ -0,0 +1,60 @@
+import org.wordpress.android.util.signature.SignatureUtils
+import javax.inject.Inject
+
+class ReaderSavedPostsProvider  : QueryContentProvider() {
```
Super small detail: I think there's an extra space here

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/reader/savedposts/provider/ReaderSavedPostsProvider.kt:
```diff
@@ -0,0 +1,60 @@
+class ReaderSavedPostsProvider  : QueryContentProvider() {
+    @Inject lateinit var queryResult: QueryResult
+    @Inject lateinit var readerPostTableWrapper: ReaderPostTableWrapper
+    @Inject lateinit var signatureUtils: SignatureUtils
```
I've extracted the common Provider logic to `ClientVerification` class (example [here](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/java/org/wordpress/android/sharedlogin/provider/SharedLoginProvider.kt#L33)). Would it make sense to use it instead?

Basically it would replace lines 37-41 with something like:
```
if (clientVerification.canTrust(callingPackage)) {
```
And `SignatureUtils` dependency would be replaced by `ClientVerification`.

develric replied:
I may have missed it after merging from trunk, makes total sense 👍

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java:
```diff
@@ -280,7 +280,10 @@ public enum UndeletablePrefKey implements PrefKey {
-        IS_FIRST_TRY_USER_FLAGS_JETPACK
+        IS_FIRST_TRY_USER_FLAGS_JETPACK,
+
+        // Indicates if this is the first time we try to get the user flags in Jetpack automatically
```
Would it make sense to replace this part of the comment: `"...to get the user flags..."` with `"...to get the reader saved posts..."` or something like that?

develric replied:
copy/paste at play 😄

RenanLukas commented on WordPress/src/test/java/org/wordpress/android/sharedlogin/resolver/SharedLoginResolverTest.kt:
```diff
@@ -106,10 +116,23 @@ class SharedLoginResolverTest {
 
     @Test
-    fun `Should dispatch UpdateTokenPayload if access token is NOT empty`() {
+    fun `Should dispatch UpdateTokenPayload if access token is NOT, flags and saved posts are migrated`() {
```
I think we might have lost a word after the `"...NOT"`. Should it be, like `"...NOT empty, flags..."` instead?

RenanLukas commented on WordPress/src/test/java/org/wordpress/android/sharedlogin/resolver/SharedLoginResolverTest.kt:
```diff
@@ -106,10 +116,23 @@ class SharedLoginResolverTest {
+    fun `Should dispatch UpdateTokenPayload if access token is NOT, flags and saved posts are migrated`() {
         featureEnabled()
+        onSuccessFlagsCaptor = argumentCaptor()
+        onSuccessRaaderPostsCaptor = argumentCaptor()
```
I think there was a typo in this variable name (`onSuccessRaaderPostsCaptor` -> `onSuccessReaderPostsCaptor`).

RenanLukas commented on WordPress/src/test/java/org/wordpress/android/sharedlogin/resolver/SharedLoginResolverTest.kt:
```diff
@@ -124,7 +147,19 @@ class SharedLoginResolverTest {
     fun `Should NOT dispatch UpdateTokenPayload if access token IS empty`() {
         featureEnabled()
+        onSuccessFlagsCaptor = argumentCaptor()
+        onSuccessRaaderPostsCaptor = argumentCaptor()
```
I think there was a typo in `"Reader"`

RenanLukas reviewed (commented):
Hey @develric , thanks for the PR

I've left a few comments. Please let me know if they make sense 👍

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/reader/savedposts/resolver/ReaderSavedPostsResolver.kt:
```diff
@@ -0,0 +1,102 @@
+                    )
+            )
+
+            readerTagTableWrapper.getBookmarkTags()!!.first().let {
```
Would it make sense to throw an explicit Exception instead of using the non-null assertion operator?
e.g.
```
readerTagTableWrapper.getBookmarkTags()?.let {
            readerPostTableWrapper.addOrUpdatePosts(it.first(), posts)
} ?: throw NullPointerException("${ReaderTagList::class.java.simpleName} not found.")
```

develric replied:
Resolving this one since it seems the same as [this one](https://github.com/wordpress-mobile/WordPress-Android/pull/17290#discussion_r992566664). Let me know if I'm wrong of course 🙇

RenanLukas replied:
It's the same. My bad, my first review was "gone" from GitHub and I thought it wasn't submitted, but apparently it was 🤷‍♂️

develric committed:
Fixing typos

develric committed:
Using ClientVerification

develric committed:
Refactoring to use requireNotNull and GenericError with message

develric commented:
Thanks for the review and suggestions @RenanLukas 🙇 , I should have followed up to your notes and ready for another pass 👍 
PS: note that I've changed a bit the logic to handle the case where there is no saved posts as a success and added this test step

>5 - Check the above is working with empty saved posts list scenario

develric committed:
Fixing detekt

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/reader/savedposts/provider/ReaderSavedPostsProvider.kt:
```diff
@@ -0,0 +1,56 @@
+        sortOrder: String?
+    ): Cursor? {
+        inject()
+        return context?.let {
```
`JetpackProviderSyncFeatureConfig` was already merged to `trunk`. Would it make sense to add it now or we're doing it in a different PR?

RenanLukas reviewed (approved):
Thanks for the changes @develric ! Besides [this question](https://github.com/wordpress-mobile/WordPress-Android/pull/17290#discussion_r995179342) I have nothing no add.

I'm going to approve it so you can merge it in case you prefer us to handle the above in a separate PR :shipit:

develric committed:
Merge branch 'trunk' into issue/reader-saved-posts-sync

# Conflicts:
#	WordPress/build.gradle
#	WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java
#	WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefsWrapper.kt
#	libs/analytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java
#	libs/analytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java

develric committed:
Adding feature flag check on Provider.

develric commented:
Hey @RenanLukas , since I needed to merge a [few conflicting files from trunk](cc8cfdacb87e4598ae2fff08954c993c08706d05), I added the [feature flag in the provider](6d74cd9e7aa08930a811b135e2faebe91e697eee) as per your advice. Hope it's fine and you can have a quick check/test round here 🙇

RenanLukas reviewed (approved):
@develric  tested again and it works as expected 👍 :shipit:
